# An introduction to functional programming in Haskell

Derived from Prof Miles Berry by Ian Benson

(c) ATM Functional Programming and Computer Algebra Working Group, 2025  licence CC BY-NC-ND Attribution-NonCommercial-NoDerivs https://creativecommons.org/licenses/

PGCE Secondary Computing Lecture 17

### A few easy functions

`halve x = x / 2` -- this defines a function halve which divides a number by two. Try it.

If you want to show off, you can do this as an anonymous function, but don't assume anyone will undestand what you've done or why you've done it!

`halve = \x -> x/2`

In [1]:
halve x = x/2

In [2]:
halve 12 -- should give the output 6!

6.0

Create functions `double` and `square` that take a number and return the double and square of that number respectively.

Create functions `isEven` and `isOdd` that take a number and return True if the number is even or odd respectively. Know that `` `mod` `` performs modular arithmetic, so `` 5 `mod` 2 `` returns 1.

Create a function add which takes two inputs and returns their sum.

Check that these work with appropriate tests. 

Create a test to return `True` if a number $n$ is a divisible by a given number $m$.

In [6]:
factors n = [x | x <- [1.. abs n], n `mod` x == 0]

Create `factors` that takes a number and returns a list of the factors of that number. 

You can return a list comprehension which is the Haskell equivalent of a set generator $ s = \{ 2x \;|\; x \in \mathbb{Z}, x \ge 2, x \bmod 2 \neq 0 \} $ that is `s = [2*x | x <- integers, isOdd x] -- an equivalent Haskell list comprehension where integers = [2..]`

Create `isPrime` that takes a number and returns True if the number is prime and False otherwise. You can use the `length` function to test the length of a list.

In [8]:
isPrime n = length ( factors n) == 2

In [9]:
factors 24

[1,2,3,4,6,8,12,24]

In [10]:
isPrime 23

True

In [13]:
isPrime 1001
factors 1001

False

[1,7,11,13,77,91,143,1001]

[1,5,7,13,29,35,65,91,145,203,377,455,1015,1885,2639,13195]

## List functions

Write functions `head'` and `tail'` that take a list and return the first and all the following elements respectively. 

For example:

`head'([3,1,4,1,5,9]) == 3`  
`tail'([3,1,4,1,5,9]) == [1,4,1,5,9]`

In [ ]:
head' 

In [ ]:
head' [1..4]

In [ ]:
tail' 

In [ ]:
tail' [1..4]

## Some recursive functions

Create a function `sum'` which accepts a list as input and returns the sum of the elements of the list.

(Harder) Create functions `bin2dec` and `dec2bin` which convert between binary and decimal numbers. The binary numbers should be represented as strings. You need to know that `read` converts a numerical string to a number, that `show` converts a number to a string and that `++` concatenates strings.

These should be recursive functions which call themselves; avoid using for or while loops, and do not use built in functions!

Write a recursive function `len'` which returns the number of items in a list.

Write a recursive function `reverse'` which takes a list and returns the list in reverse order.

You could use the `head'` and `tail'` functions you have already written.

In [ ]:
sum' 

In [ ]:
sum' [1..5]

In [ ]:
bin2dec

In [ ]:
bin2dec "1010"

In [ ]:
dec2bin 

In [ ]:
dec2bin 42

In [ ]:
len' 

In [ ]:
len' "hello world"

In [ ]:
reverse' 

In [ ]:
reverse' "miles"

### An important algorithm

What does this do?

In [ ]:
gcf (a, 0) = a
gcf (a, b) = gcf (b, a `mod` b)

In [ ]:
gcf (120, 24)

Use this to create a function `lcm`

In [ ]:
lcm (a, b) 

In [ ]:
lcm (12, 15)

## Higher order functions

Create a function `map'` which takes a function and a list as input and returns a list of the results of applying the function to each element of the list.

Create a function `filter'` which takes a Boolean function and a list as input and returns a list of the elements for which the function returns `True`.

Create a function `reduce` which takes a function and a list as input and returns the result of applying the function to each element of the list in turn, starting with the first two elements of the list.

It would be nice to do these as recursive functions. 

In [ ]:
map' f 

In [ ]:
map' halve [2,6..14]

In [ ]:
filter' f 

In [ ]:
filter' even [1..10]

In [ ]:
reduce f 

In [ ]:
reduce (*) [1..4]

## Combining functions

Use `map'` and `square` to list the squares of the first 10 positive integers.

Use `filter'` and `isEven` to list the even numbers in the list [1,2,3,4,5,6,7,8,9,10].

Use `reduce` and `add` to sum the first 10 positive integers.

Use `filter'` and `map'` to find even numbers in the squares of the first 10 positive integers.

use `map'` and `reduce` to find the sum of the squares of the first 10 positive integers.

use `filter'` and `isPrime` to find the prime numbers below 100.

use `factors` and `filter'` to create a function which finds the prime factors of a number.

## Quicksort

The quicksort algorithm partitions a list into those numbers below a pivot, the pivot (typically the first item on the original list), and those numbers above the pivot. It then recursively sorts the two sublists.

Use `filter` and list concatenation (`++`) to create a function quicksort which sorts a list of numbers.

In [ ]:
quicksort 

In [ ]:
quicksort [3,1,4,1,5,9,2,7]

## Project Euler questions

If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.

Find the sum of all the multiples of 3 or 5 below 1000.

What is the largest prime factor of 13195?